spring-integration-ftp/src/main/java/org/springframework/integration/ftp/session/FtpSession.java

```diff
@@ -200,18 +200,35 @@ public boolean rmdir(String directory) throws IOException {
 	public boolean exists(String path) throws IOException {
 		Assert.hasText(path, "'path' must not be empty");
 
-		String currentWorkingPath = this.client.printWorkingDirectory();
-		Assert.state(currentWorkingPath != null,
-				"working directory cannot be determined, therefore exists check can not be completed");
 		boolean exists = false;
-
+		InputStream inputStream = null;
 		try {
-			if (this.client.changeWorkingDirectory(path)) {
-				exists = true;
-			}
+			inputStream = this.client.retrieveFileStream(path);
```

Comment:
No; my fault; I should have been clearer - sorry.

We could also add new methods `fileExists` and `dirExists` - to avoid always doing the fallback - and use the `fileExists` method for our Fail processing.

This "directory exists" problem is only for people using the template to check for directories being present - the framework uses session.exists() for that purpose.